libsolidity/inlineasm/AsmCodeGen.cpp

```diff
@@ -0,0 +1,263 @@
+	):
+		m_state(_state)
+	{
+		if (_identifierAccess)
+			m_identifierAccess = _identifierAccess;
+		else
+			m_identifierAccess = [](assembly::Identifier const&, eth::Assembly&, CodeGenerator::IdentifierContext) { return false; };
+	}
+
+	void operator()(Instruction const& _instruction)
+	{
+		m_state.assembly.append(_instruction.instruction);
+	}
+	void operator()(assembly::Literal const& _literal)
+	{
+		if (_literal.isNumber)
+			m_state.assembly.append(u256(_literal.value));
+		else if (_literal.value.size() > 32)
+			m_state.addError(
+				Error::Type::TypeError,
+				"String literal too long (" + boost::lexical_cast<string>(_literal.value.size()) + " > 32)"
+			);
+		else
+			m_state.assembly.append(_literal.value);
+	}
+	void operator()(assembly::Identifier const& _identifier)
+	{
+		// First search local variables, then labels, then externals.
+		if (int const* stackHeight = m_state.findVariable(_identifier.name))
+		{
+			int heightDiff = m_state.assembly.deposit() - *stackHeight;
+			if (heightDiff <= 0 || heightDiff > 16)
+				//@TODO location
+				m_state.addError(
+					Error::Type::TypeError,
+					"Variable inaccessible, too deep inside stack (" + boost::lexical_cast<string>(heightDiff) + ")"
+				);
+			else
+				m_state.assembly.append(eth::dupInstruction(heightDiff));
+			return;
+		}
+		else if (eth::AssemblyItem const* label = m_state.findLabel(_identifier.name))
+			m_state.assembly.append(label->pushTag());
+		else if (!m_identifierAccess(_identifier, m_state.assembly, CodeGenerator::IdentifierContext::RValue))
+			m_state.addError(
+				Error::Type::DeclarationError,
+				"Identifier \"" + string(_identifier.name) + "\" not found or not unique"
+			);
+	}
+	void operator()(FunctionalInstruction const& _instr)
+	{
+		for (auto it = _instr.arguments.rbegin(); it != _instr.arguments.rend(); ++it)
+		{
+			int height = m_state.assembly.deposit();
+			boost::apply_visitor(*this, *it);
+			expectDeposit(1, height);
+		}
+		(*this)(_instr.instruction);
+	}
+	void operator()(Label const& _label)
+	{
+		m_state.assembly.append(m_state.labels.at(_label.name));
+	}
+	void operator()(assembly::Assignment const& _assignment)
+	{
+		generateAssignment(_assignment.variableName);
+	}
+	void operator()(FunctionalAssignment const& _assignment)
+	{
+		int height = m_state.assembly.deposit();
+		boost::apply_visitor(*this, *_assignment.value);
+		expectDeposit(1, height);
+		generateAssignment(_assignment.variableName);
+	}
+	void operator()(assembly::VariableDeclaration const& _varDecl)
+	{
+		int height = m_state.assembly.deposit();
+		boost::apply_visitor(*this, *_varDecl.value);
+		expectDeposit(1, height);
+		m_state.variables.push_back(make_pair(_varDecl.name, height));
+	}
+	void operator()(assembly::Block const& _block)
+	{
+		size_t numVariables = m_state.variables.size();
+		std::for_each(_block.statements.begin(), _block.statements.end(), boost::apply_visitor(*this));
+		// pop variables
+		//@TODO check height before and after
```

Comment:
I opted for not checking the height on purpose - we could make that a bit clearer in the documentation.